### PR TITLE
Don't force editor composition during ASL

### PR DESCRIPTION
**Pull request by rchande**
Hack_ThemeColorFixer was running during ASL for some users and forcing the editor to compose at that time. Instead, we wait until a text editor of the appropriate content type has been opened.

Addresses devdiv bug https://devdiv.visualstudio.com/DevDiv/_workitems/edit/130129

If you take a look at the [stacks](http://perfwatson/#?q=f:microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.HACK_ThemeColorFixer.RefreshThemeColors%20v%3D15*&t&s=microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.HACK_ThemeColorFixer.RefreshThemeColors&sf=Stacks), you'll see that the "scenarios" this occurs in are almost entirely all async solution load and package load and that the delay time is MEF composition. That suggests to me that the delay is not a mid-session settings sync but simply us loading the editor extra early to start copying the settings.

Tag @dotnet/roslyn-ide

### CODE REVIEW

File: src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs [hunk 1]
```diff
@@ -16,8 +21,10 @@ namespace Microsoft.VisualStudio.LanguageServices
     /// from its default blue. As a work around, we listen to <see cref="IClassificationFormatMap.ClassificationFormatMappingChanged"/>
     /// and update the classification format maps that we care about.
     /// </summary>
-    [Export]
-    internal sealed class HACK_ThemeColorFixer
+    [Export(typeof(IWpfTextViewConnectionListener))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Interactive)]
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs]: Why interactive? Is that required for any of the features where we need this?

Comment (rchande) [inline src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs]: The docs [here](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.text.editor.iwpftextviewconnectionlistener.aspx) seem to imply that _some_ TextViewRole must be specified

Maybe Analyzable would be better? I'm not really sure. @olegtk, any advice?

File: src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs [hunk 2]
```diff
@@ -116,5 +122,29 @@ public void RefreshThemeColors()
...
+            // it was possible for it to run before any files had been opened.This meant
+            // that our requests for colors and editor services would cause the composition of
+            // the editor, thus leading to a UI delay during ASL.
+            // To avoid this, we now wait until a text editor has been opened to schedule the
+            // color synchronization.
+
+            if (!_done)
+            {
+                _done = true;
+                // Run on the UI thread when VS is idle
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs]: Comment seems obvious. :-)

File: src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs [hunk 3]
```diff
@@ -116,5 +122,29 @@ public void RefreshThemeColors()
...
+        public void SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            // DevDiv https://devdiv.visualstudio.com/DevDiv/_workitems/edit/130129:
+            //
+            // We used to schedule the RefreshThemeColors task when the Roslyn Package loaded
+            // during async solution load. Even though we scheduled the task for UI thread idle,
+            // it was possible for it to run before any files had been opened.This meant
+            // that our requests for colors and editor services would cause the composition of
+            // the editor, thus leading to a UI delay during ASL.
+            // To avoid this, we now wait until a text editor has been opened to schedule the
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs]: Not sure the history story in this comment makes sense...shouldn't this have been where it was all along?

**Comment by MattGertz** [comment]
Mondo escrow triage:  keeping this, but please make sure that the wins are clearly defined.  If not a regression, the bar is a 20% or greater improvement in the appropriate perf metric.

**Comment by rchande** [comment]
Tag @MattGertz for approval. This fix should completely remove this UI delay--a 100% improvement!

**Customer scenario**
Customer opens a solution containing C# or VB projects. They experience a UI delay during async solution load.
**Bugs this fixes:** https://devdiv.visualstudio.com/DevDiv/_workitems/edit/130129
**Workarounds, if any**: None
**Risk**: Low: we move an operation that required the VS Editor to be MEF composed to after the editor is composed.
**Performance impact**: Positive: we won't force composition of the editor as soon as the Roslyn package is loaded.
**Is this a regression from a previous update?**: No, we've had this issue for as long as Roslyn has had colorized tooltips.
**Root cause analysis:**: We didn't realize this would cause UI delays when the code was written (years ago).
**How was the bug found?**: Telemetry

**Comment by MattGertz** [comment]
@rchande Very glad to see this!  But, alas, we just entered Escrow a few minutes ago and the approval process is different.  Please file a VSO bug w/PU Triage = Submit (and ASK template copied there as well).  I'll review all-up & send to JoC on Monday.

**Comment by rchande** [comment]
@MattGertz Alas. Added template to VSO bug and marked for submit.

File: src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs [hunk 4]
```diff
@@ -16,12 +21,16 @@ namespace Microsoft.VisualStudio.LanguageServices
     /// from its default blue. As a work around, we listen to <see cref="IClassificationFormatMap.ClassificationFormatMappingChanged"/>
     /// and update the classification format maps that we care about.
     /// </summary>
-    [Export]
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/HACK_ThemeColorFixer.cs]: It appears this exact type is also being consumed [here](https://github.com/dotnet/roslyn/blob/be3834ceddde9ac0ceb5921cddfc51f46965d6a2/src/VisualStudio/InteractiveServices/Interactive/VsInteractiveWindowPackage.cs#L48) which is being caught by the failing integration tests.

**Comment by jasonmalinowski** [comment]
Integration tests are failing, and it looks at first glance that the failure is legit: the dump shows a failure trying to load the interactive package, and I think I see a problem.

**Comment by rchande** [comment]
test windows_debug_vs-integration_prtest please

**Comment by rchande** [comment]
Ping @jasonmalinowski -- removed Interactive consumption of this type and verified that color changes are propagated into C# interactive windows. Can you sign off?

**Comment by rchande** [comment]
VSI Interactive Window tests were failing because the InteractivePackage was trying to import ColorThemeFixer which isn't actually exported any more. Since it's exported as an ITextViewConnectionListener, the editor takes care of composing it and calling it at the right time. I verified that color changes are propgated into the interactive window.

**Pull request merged**